File: Euler-Particles/src/shape/Star.java
package shape;

import processing.core.PApplet;
import processing.core.PVector;

public class Star extends Polygon {

    public Star(PApplet p) {
        super(p);
        int points = (int)p.random(4, 12);
        initVertices(points);
    }

    public Star(PApplet p, int points) {
        super(p, points);
    }

    protected void initVertices(int points) {
        float theta = 0;
        PApplet p = getPApplet();
        points = PApplet.constrain(points, 3, 20);
        PVector[] vertices = new PVector[points * 2];
        float deltaTheta = p.TWO_PI / (points * 2);

        for (int i = 0; i < vertices.length; i++) {
            float r = getRadius();

            if (i % 2 == 0) {
                r /= 2;
            }

            float x = PApplet.cos(theta) * r;
            float y = PApplet.sin(theta) * r;
            theta += deltaTheta;
            vertices[i] = new PVector(x, y);
        }

        setVertices(vertices);
    }
}

File: docs/Euler-Particles/shape/star.js
class Star extends Polygon {

    constructor(points) {
        super(points);
    }

    _initVertices(points) {
        let theta = 0;
        points = constrain(points, 3, 20);
        let deltaTheta = TWO_PI / (points * 2);

        for (let i = 0; i < points * 2; i++) {
            let r = this.radius;

            if (i % 2 == 0) {
                r /= 2;
            }

            let x = cos(theta) * r;
            let y = sin(theta) * r;
            theta += deltaTheta;
            this.vertices.push(createVector(x, y));
        }
    }
}
